File: yucai/server/internal/holding/domain/brent.go
package domain

import (
	"errors"
	"math"
)

// brentMaxIter 是 Brent 迭代上限。bracket 内连续函数理论上远早于此收敛,
// 超限视为数值异常(fail-closed 返错,不返回伪根)。
const brentMaxIter = 100

var errBrentNoConvergence = errors.New("brent: max iterations exceeded")

// brentRoot 在 [lo, hi] 内用 Brent(1973)方法求 f 的根。
// 语义对齐 scipy.optimize.brentq:逆二次插值 + secant,bisection 兜底,
// 收敛判据 |x−x₀| ≤ xtol + rtol·|x|(scipy 缺省 xtol=2e-12, rtol=4·eps)。
// 前置:fa=f(lo)、fb=f(hi) 由调用方预传,且 fa·fb ≤ 0(已反号);
// 端点恰为零直接返回端点;同号返错。
func brentRoot(f func(float64) float64, lo, hi, fa, fb, xtol, rtol float64) (float64, error) {
	if fa == 0 {
		return lo, nil
	}
	if fb == 0 {
		return hi, nil
	}
	if fa*fb > 0 {
		return 0, errors.New("brent: root not bracketed")
	}

	a, b := lo, hi
	c, fc := a, fa
	d, e := b-a, b-a // d=上一步位移, e=上上步位移(插值步判断用)

	for iter := 0; iter < brentMaxIter; iter++ {
		if fb*fc > 0 {
			c, fc = a, fa
			d, e = b-a, b-a
		}
		if math.Abs(fc) < math.Abs(fb) {
			a, b, c = b, c, b
			fa, fb, fc = fb, fc, fb
		}
		tol1 := 2*rtol*math.Abs(b) + 0.5*xtol
		xm := 0.5 * (c - b)
		if math.Abs(xm) <= tol1 || fb == 0 {
			return b, nil
		}
		if math.Abs(e) >= tol1 && math.Abs(fa) > math.Abs(fb) {
			s := fb / fa
			var p, q float64
			if a == c {
				// secant
				p = 2 * xm * s
				q = 1 - s
			} else {
				// 逆二次插值
				q = fa / fc
				r := fb / fc
				p = s * (2*xm*q*(q-r) - (b-a)*(r-1))
				q = (q - 1) * (r - 1) * (s - 1)
			}
			if p > 0 {
				q = -q
			}
			p = math.Abs(p)
			min1 := 3*xm*q - math.Abs(tol1*q)
			min2 := math.Abs(e * q)
			if 2*p < math.Min(min1, min2) {
				e = d
				d = p / q
			} else {
				d = xm
				e = d
			}
		} else {
			d = xm
			e = d
		}
		a, fa = b, fb
		if math.Abs(d) > tol1 {
			b += d
		} else {
			b += math.Copysign(tol1, xm)
		}
		fb = f(b)
	}
	return 0, errBrentNoConvergence
}

File: yucai/client/lib/holding/domain/return_engine/brent.dart
import 'dart:math' as math;

/// Brent(1973)求根 — R5-G brent.go 的 Dart 逐函数移植(设计/语义对齐
/// scipy.optimize.brentq;包络内保证收敛)。null=不收敛/未括根。
/// oracle:test/holding/domain/return_engine/(G 用例移植)。

const int brentMaxIter = 100;

double? brentRoot(
  double Function(double) f,
  double lo,
  double hi,
  double fa,
  double fb,
  double xtol,
  double rtol,
) {
  if (fa == 0) return lo;
  if (fb == 0) return hi;
  if (fa * fb > 0) return null;

  var a = lo, b = hi;
  var c = a, fc = fa;
  var d = b - a, e = b - a;

  for (var iter = 0; iter < brentMaxIter; iter++) {
    if (fb * fc > 0) {
      c = a;
      fc = fa;
      d = b - a;
      e = b - a;
    }
    if (fc.abs() < fb.abs()) {
      a = b; b = c; c = a;
      fa = fb; fb = fc; fc = fa;
    }
    final tol1 = 2 * rtol * b.abs() + 0.5 * xtol;
    final xm = 0.5 * (c - b);
    if (xm.abs() <= tol1 || fb == 0) return b;
    if (e.abs() >= tol1 && fa.abs() > fb.abs()) {
      final s = fb / fa;
      double p, q;
      if (a == c) {
        p = 2 * xm * s;
        q = 1 - s;
      } else {
        q = fa / fc;
        final r = fb / fc;
        p = s * (2 * xm * q * (q - r) - (b - a) * (r - 1));
        q = (q - 1) * (r - 1) * (s - 1);
      }
      if (p > 0) q = -q;
      p = p.abs();
      final min1 = 3 * xm * q - (tol1 * q).abs();
      final min2 = (e * q).abs();
      if (2 * p < math.min(min1, min2)) {
        e = d;
        d = p / q;
      } else {
        d = xm;
        e = d;
      }
    } else {
      d = xm;
      e = d;
    }
    a = b;
    fa = fb;
    if (d.abs() > tol1) {
      b += d;
    } else {
      b += _copysign(tol1, xm);
    }
    fb = f(b);
  }
  return null;
}

double _copysign(double v, double ref) => ref.isNegative ? -v.abs() : v.abs();
